File: dashboard/src/lib/pnl.ts
/**
 * Direction-aware P&L utilities.
 *
 * Single source of truth for classifying winners vs losers on the Signals
 * dashboard. Mirrors `python_agents/utils/pnl.py`; keep the two in sync.
 */

export type SignalDirectionLike = string | null | undefined;

export interface SignalLike {
  direction?: SignalDirectionLike;
  entry_price?: number | string | null;
  exit_price?: number | string | null;
  current_price?: number | string | null;
  price?: number | string | null;
  status?: string | null;
  metadata?: Record<string, any> | null;
}

export type OutcomeBucket = "profit" | "loss" | "pending";

function coerceFloat(value: unknown): number | null {
  if (value === null || value === undefined || value === "") return null;
  const n = Number(value);
  return Number.isFinite(n) ? n : null;
}

export function normalizeDirection(raw: SignalDirectionLike | unknown): "long" | "short" {
  if (raw === null || raw === undefined) return "long";
  const text = String(raw).trim().toLowerCase();
  if (text === "short" || text === "sell" || text === "down" || text === "bear") return "short";
  return "long";
}

/**
 * Returns the signed realized P&L percentage (positive = profit), or `null`
 * when inputs are insufficient.
 */
export function computeSignalPnlPct(signal: SignalLike): number | null {
  const meta = signal.metadata ?? {};
  const entry = coerceFloat(signal.entry_price ?? meta.entry_price ?? signal.price);
  if (entry === null || entry <= 0) return null;

  const ref = coerceFloat(
    signal.exit_price ?? meta.exit_price ?? signal.current_price ?? meta.current_price_at_signal,
  );
  if (ref === null) return null;

  const direction = normalizeDirection(
    signal.direction ?? meta.direction ?? meta.position_bias,
  );

  const raw = ((ref - entry) / entry) * 100;
  return direction === "short" ? -raw : raw;
}

export function isProfitable(signal: SignalLike): boolean {
  const p = computeSignalPnlPct(signal);
  return p !== null && p > 0;
}

export function classifySignalOutcome(signal: SignalLike): OutcomeBucket {
  const p = computeSignalPnlPct(signal);
  if (p === null) return "pending";
  if (p > 0) return "profit";
  if (p < 0) return "loss";
  return "pending";
}

/**
 * Resolve the best realized reference price from a signal blob that may also
 * carry horizon outcomes. Used by the outcome cards which need to show the
 * hit / close price alongside entry.
 */
export function resolveRealizedPrice(signal: any): number | null {
  const meta = signal?.metadata ?? {};
  const horizons: any[] = Array.isArray(meta.target_horizons) ? meta.target_horizons : [];
  const hit = horizons.find((h) => h?.status === "hit");
  const primary = hit || horizons.find((h) => ["missed", "expired"].includes(String(h?.status)));
  const candidates = [
    signal?.exit_price,
    meta.exit_price,
    primary?.actual_price,
    signal?.current_price,
    meta.current_price_at_signal,
  ];
  for (const c of candidates) {
    const num = coerceFloat(c);
    if (num !== null && num > 0) return num;
  }
  return null;
}

File: python_agents/utils/pnl.py
"""Direction-aware P&L utilities.

Single source of truth for converting raw entry/exit prices into a signed
P&L percentage that honors the trade direction (long vs short).

The dashboard mirrors this logic in ``dashboard/src/lib/pnl.ts``; keep the
two in sync.
"""
from __future__ import annotations

from typing import Any, Literal, Mapping, Optional

OutcomeBucket = Literal["profit", "loss", "pending"]


def _coerce_direction(raw: Any) -> str:
    """Normalize a direction-ish value to ``'long'`` or ``'short'``."""
    if raw is None:
        return "long"
    text = str(raw).strip().lower()
    if text in ("short", "sell", "down", "bear"):
        return "short"
    return "long"


def _coerce_float(value: Any) -> Optional[float]:
    try:
        if value is None:
            return None
        num = float(value)
    except (TypeError, ValueError):
        return None
    if num != num:  # NaN guard
        return None
    return num


def compute_signal_pnl_pct(signal: Mapping[str, Any]) -> Optional[float]:
    """Return the direction-adjusted realized P&L percentage.

    ``signal`` may use any of the following keys:
      - ``direction``: ``'long'`` or ``'short'`` (falls back to ``position_bias``
        inside ``metadata`` when absent).
      - ``entry_price``: entry price; falls back to ``price`` and
        ``metadata.entry_price``.
      - ``exit_price`` (preferred) or ``current_price``: realized or latest
        reference price. Also reads ``metadata.exit_price`` and
        ``metadata.current_price_at_signal`` as last-resort fallbacks.

    Returns ``None`` when either entry or reference price is missing,
    otherwise the signed return in percent (positive = profit).
    """
    metadata = signal.get("metadata") if isinstance(signal, Mapping) else None
    if not isinstance(metadata, Mapping):
        metadata = {}

    entry = _coerce_float(
        signal.get("entry_price")
        or metadata.get("entry_price")
        or signal.get("price")
    )
    if entry is None or entry <= 0:
        return None

    ref = _coerce_float(
        signal.get("exit_price")
        or metadata.get("exit_price")
        or signal.get("current_price")
        or metadata.get("current_price_at_signal")
    )
    if ref is None:
        return None

    direction = _coerce_direction(
        signal.get("direction") or metadata.get("direction") or metadata.get("position_bias")
    )

    raw_pct = ((ref - entry) / entry) * 100.0
    return -raw_pct if direction == "short" else raw_pct


def classify_signal_outcome(signal: Mapping[str, Any]) -> OutcomeBucket:
    """Bucket a signal into ``'profit'``, ``'loss'`` or ``'pending'``.

    ``'pending'`` is returned when we cannot compute a realized P&L yet
    (missing exit/current price or zero entry).
    """
    pnl = compute_signal_pnl_pct(signal)
    if pnl is None:
        return "pending"
    if pnl > 0:
        return "profit"
    if pnl < 0:
        return "loss"
    return "pending"


def is_profitable(signal: Mapping[str, Any]) -> bool:
    """Convenience wrapper: ``True`` when the signal realized a positive P&L."""
    pnl = compute_signal_pnl_pct(signal)
    return pnl is not None and pnl > 0
